src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -102,8 +147,49 @@ public void RequestHeartbeat()
             }
         }
 
+        // private methods
+        private CommandWireProtocol<BsonDocument> InitializeIsMasterProtocol(IConnection connection)
+        {
+            BsonDocument isMasterCommand;
+            var commandResponseHandling = CommandResponseHandling.Return;
+            if (connection.Description.IsMasterResult.TopologyVersion != null)
+            {
+                // the final Readtimeout will be:
+                // connectTimeoutMS (the default heartbeat read timeout) + maxAwaitTimeMS (_heartbeatInterval)
+                connection.AddAdditionalReadTimeout(_heartbeatInterval);
+                commandResponseHandling = CommandResponseHandling.ExhaustAllowed;
+
+                isMasterCommand = IsMasterHelper.CreateCommand(connection.Description.IsMasterResult.TopologyVersion, _heartbeatInterval);
+            }
+            else
+            {
+                isMasterCommand = IsMasterHelper.CreateCommand();
+            }
+
+            return IsMasterHelper.CreateProtocol(isMasterCommand, commandResponseHandling);
+        }
+
+        private async Task<IsMasterResult> InitializeConnectionAsync(CancellationToken cancellationToken)
+        {
+            lock (_lock)
+            {
+                _connection = _connectionFactory.CreateConnection(_serverId, _endPoint);
+            }
+            // if we are cancelling, it's because the server has
+            // been shut down and we really don't need to wait.
+            var stopwatch = Stopwatch.StartNew();
+            await _connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+            stopwatch.Stop();
+
+            _handshakeBuildInfoResult = _connection.Description.BuildInfoResult;
+            _roundTripTimeMonitor.AddSample(stopwatch.Elapsed);
+            return _connection.Description.IsMasterResult;
+        }
+
         private async Task MonitorServerAsync()
         {
+            await Task.Yield(); // return control immediately
```

Comment:
Similar to before: requesting a bit of additional context as to why we need this `Task.Yield()` :)